src/io/parquet/write/mod.rs

```diff
@@ -163,8 +168,29 @@ pub fn array_to_pages(
                     )
                 })
             }
-            _ => array_to_page(array, type_, nested, options, encoding)
-                .map(|page| DynIter::new(std::iter::once(Ok(page)))),
+            _ => {
+                const DEFAULT_PAGE_SIZE: usize = 1024 * 1024;
+                let page_size = options.data_pagesize_limit.unwrap_or(DEFAULT_PAGE_SIZE);
+                let bytes_per_row =
+                    (((array_byte_size as f64) / (array.len() as f64)) as usize).max(1);
+                let rows_per_page = (page_size / bytes_per_row).max(1);
+
+                let vs: Vec<Result<EncodedPage>> = (0..array.len())
+                    .step_by(rows_per_page)
+                    .map(|offset| {
+                        let length = if offset + rows_per_page > array.len() {
+                            array.len() - offset
+                        } else {
+                            rows_per_page
+                        };
+
+                        let sub_array = array.slice(offset, length);
+                        array_to_page(sub_array.as_ref(), type_.clone(), nested, options, encoding)
+                    })
+                    .collect();
```

Comment:
I don't think we need to collect here. We could `Box` the iterator as the `DynIter` is defined as:

```rust
pub struct DynIter<'a, V> {
    iter: Box<dyn Iterator<Item = V> + 'a + Send + Sync>,
}
```

Reply:
I tried to use
```
let iter = (0..array.len())
                    .step_by(rows_per_page)
                    .map(|offset| {
                        let length = if offset + rows_per_page > array.len() {
                            array.len() - offset
                        } else {
                            rows_per_page
                        };

                        let sub_array = array.slice(offset, length);
                        array_to_page(sub_array.as_ref(), type_.clone(), nested, options, encoding)
                    });

                Ok(DynIter::new(Box::new(iter)))
```

But got
```
   --> src/io/parquet/write/mod.rs:191:17
    |
138 |     nested: &[Nested],
    |     ------ has type `&[pages::Nested<'3>]`
...
191 |                 Ok(DynIter::new(Box::new(iter)))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'3` must outlive `'static`

```

Any advices?

Reply:
And if we `move` the borrowed items into the closure?

Reply:
Same error:

```
   Compiling arrow2 v0.14.2 (/home/sundy/work/arrow2)
error: lifetime may not live long enough
   --> src/io/parquet/write/mod.rs:191:17
    |
136 |     array: &dyn Array,
    |            - let's call the lifetime of this reference `'1`
...
191 |                 Ok(DynIter::new(Box::new(iter)))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'static`

error: lifetime may not live long enough
   --> src/io/parquet/write/mod.rs:191:17
    |
138 |     nested: &[Nested],
    |             - let's call the lifetime of this reference `'2`
...
191 |                 Ok(DynIter::new(Box::new(iter)))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'2` must outlive `'static`

error: lifetime may not live long enough
   --> src/io/parquet/write/mod.rs:191:17
    |
138 |     nested: &[Nested],
    |     ------ has type `&[pages::Nested<'3>]`
...
191 |                 Ok(DynIter::new(Box::new(iter)))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'3` must outlive `'static`

error: could not compile `arrow2` due to 3 previous errors
```

Reply:
I checked locally and I think making that work is not worth it. We need to Box the `dyn Array` and clone the `&[Nested]` slice to do so. I think collecting the iterator is better.